gammapy/maps/_sparse.pyx

```diff
@@ -0,0 +1,207 @@
+    ``ilo`` and ``ihi``.  This will return the index of the matching
+    element or the index at which ``idx`` would be inserted to
+    maintain sorted order.
+
+    Parameters
+    ----------
+    arr_in : `~numpy.ndarray`
+        Sorted array of integers.
+
+    idx : int
+        Value to be searched for.
+
+    """
+    
+    cdef int mid = 0
+    cdef int midval = 0
+
+    while(ilo < ihi):
+        mid = (ilo + ihi) // 2
+        midval = arr_in[mid]
+
+        if midval < idx:
+            ilo = mid + 1
+        elif midval > idx:
+            ihi = mid
+        else:
+            return mid
+
+    return ilo
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def binary_search2(np.ndarray[np.int64_t, ndim=1] arr_in, np.int64_t idx,
+                   np.int64_t ilo, np.int64_t ihi):
+
+    cdef int mid = 0
+    cdef int midval = 0
+
+    while(ilo < ihi):
+        mid = (ilo + ihi) // 2
+        midval = arr_in[mid]
+
+        if midval < idx:
+            ilo = mid + 1
+        elif midval > idx:
+            ihi = mid
+        else:
+            return mid, 1
+
+    return ilo, 0
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def find_in_array(np.ndarray[np.int64_t, ndim=1] idx0,
+                  np.ndarray[np.int64_t, ndim=1] idx1):
+    """Find the values in ``idx0`` contained in ``idx1``.
+
+    Parameters
+    ----------
+    idx0 : `~np.ndarray`
+        Array of unsorted input values.
+
+    idx1 : `~np.ndarray`
+        Array of sorted values to be searched. 
+
+    Returns
+    -------
+    idx : `~np.ndarray`
+        Array of indices with length of ``idx0`` with the position of
+        the given element in ``idx1``.
+
+    msk : `~np.ndarray`
+        Boolean mask with length of ``idx0`` indicating whether a
+        given value was found in ``idx1``.
+
+    """
+
+    cdef int ni = idx0.shape[0]
+    cdef int nj = idx1.shape[0]
+
+    cdef np.ndarray[np.int64_t, ndim= 1] idx_sort = np.argsort(idx0)
+    cdef int i = 0
+    cdef int ii = 0
+    cdef np.int64_t j = binary_search(idx1, idx0[idx_sort[0]], 0, nj)
+    cdef np.ndarray[np.int64_t, ndim = 1] out = np.zeros([ni], dtype=np.int64)
+    cdef np.ndarray[np.uint8_t, ndim = 1] msk = np.zeros([ni], dtype=np.uint8)
+
+    while(i < ni and j < nj):
+
+        ii = idx_sort[i]
+
+        # Current element is less
+        if idx0[ii] < idx1[j]:
+            out[ii] = j
+            i += 1
+        elif idx0[ii] > idx1[j]:
+            j += 1
+        else:
+            out[ii] = j
+            msk[ii] = True
+            i += 1
+
+    return out, msk.astype(bool)
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def merge_sparse_arrays(np.ndarray[np.int64_t, ndim=1] idx0,
+                        np.ndarray[np.float64_t, ndim=1] val0,
+                        np.ndarray[np.int64_t, ndim=1] idx1,
+                        np.ndarray[np.float64_t, ndim=1] val1,
+                        np.int64_t fill=False
```

Comment:
Cosmetic change: The `fill` option can be of type `bool`, no?
It's currently `np.int64_t`, but you use it as a bool, and the docstring says it's type `bool`.

Reply:
I was using this because I didn't know the correct type for bools in cython.  Changed now to `bint` which is an integer type that gets automatically mapped to python bool.